pkg/virt-api/rest/subresource.go

```diff
@@ -787,6 +796,40 @@ func (app *SubresourceAPIApp) UnfreezeVMIRequestHandler(request *restful.Request
 
 }
 
+func (app *SubresourceAPIApp) ResetVMIRequestHandler(request *restful.Request, response *restful.Response) {
+
+	// The VMI must be assigned to a node for us to forward
+	// the request to virt-handler. Forward the
+	preValidate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if vmi.IsFinal() {
+			// VMI is in a terminated state, with no possiblity of returning.
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("Unable to reset terminated VMI"))
+		} else if !vmi.IsScheduled() && !vmi.IsRunning() {
+			// VMI hasn't been assigned to a node yet, so there's no virt-handler
+			// available to pass the reset request to.
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("Unable to reset unscheduled VMI"))
+		}
+		return nil
+	}
+
+	// Post process any error responses in order to append human
+	// readable explanation for why the reset may have failed.
+	errorPostProcessing := func(vmi *v1.VirtualMachineInstance, err error) error {
+		// VMI reset request could have been sent while VMI was in the process of transitioning
+		// from scheduled to running.
+		if !vmi.IsRunning() {
+			return fmt.Errorf("Failed to reset non-running VMI with phase %s: %v", vmi.Status.Phase, err)
```

Comment:
> are you sure you'd rather users to be exposed to the longer error message with all those details instead of us cutting off the attempt at a reset when the VMI is not running

I prefer to expose as much of the call stack that is possible, in order to ease debugging by experts. To me, the only reason to cut the message short (and log the rest) is actual (not mere theoretical) security concerns.